counters.hh

```diff
@@ -253,23 +252,31 @@ protected:
 private:
     class shard_iterator {
     public:
-        using iterator_category = std::input_iterator_tag;
+        using iterator_category = std::forward_iterator_tag;
         using value_type = basic_counter_shard_view<is_mutable>;
         using difference_type = std::ptrdiff_t;
         using pointer = basic_counter_shard_view<is_mutable>*;
-        using reference = basic_counter_shard_view<is_mutable>&;
+        using reference = std::conditional_t<is_mutable == mutable_view::yes,
+            basic_counter_shard_view<is_mutable>&,
+            const basic_counter_shard_view<is_mutable>&>;
     private:
         managed_bytes_basic_view<is_mutable> _current;
         basic_counter_shard_view<is_mutable> _current_view;
         size_t _pos = 0;
     public:
+        shard_iterator() noexcept = default;
         shard_iterator(managed_bytes_basic_view<is_mutable> v, size_t offset) noexcept
             : _current(v), _current_view(_current), _pos(offset) { }
 
-        basic_counter_shard_view<is_mutable>& operator*() noexcept {
+        reference operator*() const noexcept requires (is_mutable == mutable_view::no) {
             return _current_view;
```

Comment:
If you return a reference to the iterator itself, then it's not a forward_iterator.